File: src/presentation/routes/editor.py
import re

from apiflask import APIBlueprint
from flask import jsonify, request, send_file

from src.domain.services import (
    editor_scene_analysis_service,
    editor_visual_service,
    enriched_render_service,
    render_service,
)
from src.infrastructure.ai_providers import image_search_client
from src.infrastructure.storage import project_repository
from src.presentation.schemas.editor import (
    EditorBuscarImagenInSchema,
    EditorCargarPlanQuerySchema,
    EditorProxyImagenInSchema,
)
from src.utils.logger import get_logger

logger = get_logger(__name__)

editor_bp = APIBlueprint("editor", __name__, url_prefix="/api/editor")


@editor_bp.post("/analizar")
def analizar():
    """Clasifica cada escena (tipo cinematografico, overlays, queries de imagen) via IA.
    Body libre (no schema estricto): {escenas: [{texto, imagen_file}], project_name}."""
    data = request.get_json(silent=True) or {}
    escenas = data.get("escenas", [])
    project_name = data.get("project_name", "")
    try:
        result = editor_scene_analysis_service.analizar_escenas(escenas, project_name)
        return jsonify({"escenas": result})
    except ValueError as exc:
        return jsonify({"error": str(exc)}), 400
    except Exception as exc:
        logger.exception("editor_analizar error")
        return jsonify({"error": str(exc)}), 500


@editor_bp.post("/buscar_imagen")
@editor_bp.input(EditorBuscarImagenInSchema)
def buscar_imagen(json_data):
    urls = image_search_client.search_images(
        json_data["query"],
        n=json_data["n"],
        serper_key=json_data["serper_key"],
        pexels_key=json_data["pexels_key"],
    )
    return jsonify({"urls": urls})


@editor_bp.post("/guardar_plan")
def guardar_plan():
    """Persiste el plan de edicion (escenas con metadatos). Body libre: {project_name, escenas}."""
    data = request.get_json(silent=True) or {}
    project_name = (data.get("project_name") or "").strip()
    if not project_name:
        return jsonify({"error": "Falta project_name"}), 400
    path = project_repository.write_editor_plan(project_name, data.get("escenas", []))
    return jsonify({"ok": True, "path": str(path)})


@editor_bp.get("/cargar_plan")
@editor_bp.input(EditorCargarPlanQuerySchema, location="query")
def cargar_plan(query_data):
    project = query_data["project"].strip()
    if not project:
        return jsonify({"existe": False})
    plan = project_repository.read_editor_plan(project)
    if plan is None:
        return jsonify({"existe": False})
    return jsonify({"existe": True, **plan})


@editor_bp.post("/proxy_imagen")
@editor_bp.input(EditorProxyImagenInSchema)
def proxy_imagen(json_data):
    result = image_search_client.proxy_image_b64(json_data["url"])
    if "error" in result:
        return jsonify(result), 400
    return jsonify(result)


@editor_bp.post("/render_enriquecido")
def render_enriquecido():
    """Render con efectos del editor: overlays de texto sincronizados, Ken Burns,
    split-screen, imagenes de referencia. Body libre: {project_name, escenas,
    resolucion, transicion, trans_dur, pexels_key, unsplash_key}. Reutiliza el mismo
    job registry que /api/render_inteligente -- se consulta via /api/estado/<job_id>
    y se descarga via /api/descargar_render/<job_id>."""
    data = request.get_json(silent=True) or {}
    try:
        result = enriched_render_service.start_render(
            project_name=data.get("project_name", ""),
            escenas=data.get("escenas", []),
            resolucion=data.get("resolucion", "1920x1080"),
            transicion=data.get("transicion", "xfade"),
            trans_dur=float(data.get("trans_dur", 0.6)),
            pexels_key=data.get("pexels_key", ""),
            unsplash_key=data.get("unsplash_key", ""),
        )
        return jsonify(result)
    except ValueError as exc:
        return jsonify({"error": str(exc)}), 400
    except Exception as exc:
        logger.exception("editor_render_enriquecido error")
        return jsonify({"error": str(exc)}), 500


@editor_bp.get("/proyecto/<proj_name>")
def proyecto(proj_name):
    """Datos del proyecto para el editor visual (imagenes, guion, audio, plan, timestamps)."""
    proj_name = re.sub(r"[^\w\-]", "_", proj_name.strip())
    if not proj_name:
        return jsonify({"error": "proyecto requerido"}), 400
    try:
        data = editor_visual_service.get_project_data(proj_name)
    except FileNotFoundError:
        return jsonify({"error": "proyecto no encontrado"}), 404
    resp = jsonify(data)
    resp.headers["Access-Control-Allow-Origin"] = "*"
    return resp


@editor_bp.get("/imagen/<proj_name>/<filename>")
def imagen(proj_name, filename):
    path = project_repository.resolve_safe_file(proj_name, "imagen", filename)
    if not path or not path.exists():
        return jsonify({"error": "no encontrado"}), 404
    resp = send_file(str(path))
    resp.headers["Access-Control-Allow-Origin"] = "*"
    resp.headers["Cross-Origin-Resource-Policy"] = "cross-origin"
    return resp


@editor_bp.get("/audio/<proj_name>/<filename>")
def audio(proj_name, filename):
    path = project_repository.resolve_safe_file(proj_name, "audio", filename)
    if not path or not path.exists():
        return jsonify({"error": "no encontrado"}), 404
    resp = send_file(str(path))
    resp.headers["Access-Control-Allow-Origin"] = "*"
    resp.headers["Cross-Origin-Resource-Policy"] = "cross-origin"
    return resp


@editor_bp.get("/transcribir/<proj_name>")
def transcribir(proj_name):
    """Transcribe (cascada automatica de backends) y asigna timestamps por escena,
    con el mismo algoritmo que el Render normal. Cachea en timestamps_escenas.json."""
    proj_name = re.sub(r"[^\w\-]", "_", proj_name.strip())
    if not proj_name:
        return jsonify({"error": "proyecto requerido"}), 400
    try:
        result = editor_visual_service.transcribe_project(proj_name)
    except FileNotFoundError:
        return jsonify({"error": "proyecto no encontrado"}), 404
    except ValueError as exc:
        return jsonify({"error": str(exc)}), 404
    except Exception as exc:
        logger.exception("editor_transcribir error")
        return jsonify({"error": str(exc)}), 500
    resp = jsonify(result)
    resp.headers["Access-Control-Allow-Origin"] = "*"
    return resp


@editor_bp.get("/estado/<job_id>")
def estado(job_id):
    """Polling de estado para el editor visual -- mismo job registry que /api/estado."""
    job = render_service.get_job(job_id)
    if job is None:
        return jsonify({"error": "no encontrado"}), 404
    resp = jsonify(job)
    resp.headers["Access-Control-Allow-Origin"] = "*"
    return resp

File: frontend/src/api/editor.ts
import { api } from "./client";

// ── Escenas / plan ────────────────────────────────────────────────────
export interface EditorScene {
  indice?: number;
  texto: string;
  imagen_file?: string;
  imagen_url?: string | null;
  ref_image_url?: string;
  tipo?: string;
  habilitado?: boolean;
  texto_overlay?: string;
  texto_overlay_pos?: string;
  texto_secundario?: string;
  color_accent?: string;
  split_label_1?: string;
  split_label_2?: string;
  numero_capitulo?: number | string;
  ref_label?: string;
  google_query?: string;
  [key: string]: unknown;
}

export interface TimestampEntry {
  escena: number;
  inicio: number;
  fin: number;
  texto: string;
}

export interface ProjectData {
  audio_url?: string | null;
  timestamps?: TimestampEntry[];
  escenas?: EditorScene[];
  guion?: string;
  images?: string[];
  [key: string]: unknown;
}

// ── Proyecto (carga principal del editor) ───────────────────────────
export function getProjectData(projectName: string) {
  return api
    .get<ProjectData>(`/editor/proyecto/${encodeURIComponent(projectName)}`)
    .then((r) => r.data);
}

export function editorImageUrl(projectName: string, filename: string): string {
  return `/api/editor/imagen/${encodeURIComponent(projectName)}/${encodeURIComponent(filename)}`;
}

export function editorAudioUrl(projectName: string, filename: string): string {
  return `/api/editor/audio/${encodeURIComponent(projectName)}/${encodeURIComponent(filename)}`;
}

// ── Analisis IA ──────────────────────────────────────────────────────
export interface AnalizarEscenasParams {
  escenas: Array<{ texto: string; imagen_file?: string }>;
  project_name: string;
}

export interface AnalizarEscenasResult {
  escenas?: EditorScene[];
  [key: string]: unknown;
}

export function analizarEscenas({ escenas, project_name }: AnalizarEscenasParams) {
  return api
    .post<AnalizarEscenasResult>("/editor/analizar", { escenas, project_name })
    .then((r) => r.data);
}

// ── Busqueda / proxy de imagenes ─────────────────────────────────────
export interface BuscarImagenParams {
  query: string;
  n?: number;
  serper_key?: string;
  pexels_key?: string;
}

export interface BuscarImagenResult {
  urls?: string[];
  [key: string]: unknown;
}

export function buscarImagen({ query, n = 4, serper_key = "", pexels_key = "" }: BuscarImagenParams) {
  return api
    .post<BuscarImagenResult>("/editor/buscar_imagen", { query, n, serper_key, pexels_key })
    .then((r) => r.data);
}

export interface ProxyImagenResult {
  b64?: string;
  error?: string;
  [key: string]: unknown;
}

export function proxyImagen(url: string) {
  return api.post<ProxyImagenResult>("/editor/proxy_imagen", { url }).then((r) => r.data);
}

// ── Plan de edicion ───────────────────────────────────────────────────
export interface GuardarPlanParams {
  project_name: string;
  escenas: EditorScene[];
}

export interface GuardarPlanResult {
  ok?: boolean;
  [key: string]: unknown;
}

export function guardarPlan({ project_name, escenas }: GuardarPlanParams) {
  return api
    .post<GuardarPlanResult>("/editor/guardar_plan", { project_name, escenas })
    .then((r) => r.data);
}

export interface CargarPlanResult {
  existe: boolean;
  escenas?: EditorScene[];
  [key: string]: unknown;
}

export function cargarPlan(project: string) {
  return api
    .get<CargarPlanResult>("/editor/cargar_plan", { params: { project } })
    .then((r) => r.data);
}

// ── Transcripcion ─────────────────────────────────────────────────────
export interface TranscribirSegment {
  seg_idx: number;
  start: number;
  end: number;
  text: string;
}

export interface TranscribirResult {
  segments?: TranscribirSegment[];
  source?: string;
  [key: string]: unknown;
}

export function transcribirProyecto(projectName: string) {
  return api
    .get<TranscribirResult>(`/editor/transcribir/${encodeURIComponent(projectName)}`)
    .then((r) => r.data);
}

// ── Render enriquecido ────────────────────────────────────────────────
export interface RenderEnriquecidoParams {
  project_name: string;
  escenas: EditorScene[];
  resolucion?: string;
  transicion?: string;
  trans_dur?: number;
  pexels_key?: string;
  unsplash_key?: string;
}

export interface RenderEnriquecidoResult {
  job_id?: string;
  error?: string;
  [key: string]: unknown;
}

export function renderEnriquecido({
  project_name,
  escenas,
  resolucion = "1920x1080",
  transicion = "xfade",
  trans_dur = 0.6,
  pexels_key = "",
  unsplash_key = "",
}: RenderEnriquecidoParams) {
  return api
    .post<RenderEnriquecidoResult>("/editor/render_enriquecido", {
      project_name,
      escenas,
      resolucion,
      transicion,
      trans_dur,
      pexels_key,
      unsplash_key,
    })
    .then((r) => r.data);
}

export interface EditorJobStatus {
  estado: string;
  mensaje?: string;
  progreso?: number;
  video_url?: string;
  error?: string;
  [key: string]: unknown;
}

export function getEditorJobStatus(jobId: string) {
  return api
    .get<EditorJobStatus>(`/editor/estado/${encodeURIComponent(jobId)}`)
    .then((r) => r.data);
}

export function getEditorDownloadUrl(jobId: string): string {
  return `/api/descargar_render/${jobId}`;
}
